### PR TITLE
Implement multiple paginators

### LINKED ISSUE #1731: 3.0 - Support for multiple pagination
Created by **cake_joe**, 11th Jan 2012. *(originally [Lighthouse ticket #102](http://cakephp.lighthouseapp.com/projects/42648/tickets/102))*:
- - - -

### PR DESCRIPTION
**Pull request by josegonzalez**
This change provides the initial support for multiple paginators on a given page.

To use it, you can do something like the following in your controller layer:

``` php
$users = $this->paginate($this->Users->find(), ['prefix' => 'users']);
$categories = $this->paginate($this->Categories->find(), ['prefix' => 'categories']);

$this->set(compact('users', 'categories'));
```

The in your view layer, specify the model option whenever the `PaginatorHelper` allows you to set options.

Closes #1731

### CODE REVIEW

**Comment by thisispaulcyr** [comment]
Thanks for the work on this! I'm eagerly awaiting its completion. Are there any updates on the status?

**Comment by markstory** [comment]
@thisispaulcyr These changes need tests written before they will be merged in. Would you be able/willing to help with that?

**Comment by thisispaulcyr** [comment]
@markstory Yes, I would be interested. However, I have never done it before. Is it something that can be learned in a couple of days? If so, are you able to point me in the right direction to get started?

**Comment by markstory** [comment]
The [book has a small section](http://book.cakephp.org/3.0/en/contributing/code.html) on getting started with contributing code.

**Comment by josegonzalez** [comment]
I added tests for a few things except the url generation, which I must admit I do not know quite *how* to test.

File: src/View/Helper/PaginatorHelper.php [hunk 1]
```diff
@@ -412,10 +419,16 @@ public function sort($key, $title = null, array $options = [])
 
         $sortKey = $this->sortKey($options['model']);
         $defaultModel = $this->defaultModel();
+        $model = Hash::get($options, 'model', $defaultModel);
```
Comment (ADmad) [inline src/View/Helper/PaginatorHelper.php]: Call to `Hash::get()` is not necessary, can just do `$model = $options['model'] ?: $defaultModel;`

File: src/Controller/Component/PaginatorComponent.php [hunk 2]
```diff
@@ -202,7 +203,8 @@ public function paginate($object, array $settings = [])
             'direction' => current($order),
             'limit' => $defaults['limit'] != $limit ? $limit : null,
             'sortDefault' => $sortDefault,
-            'directionDefault' => $directionDefault
+            'directionDefault' => $directionDefault,
+            'prefix' => Hash::get($options, 'prefix', null),
```
Comment (ADmad) [inline src/Controller/Component/PaginatorComponent.php]: Call to `Hash::get()` can be avoided by adding `'prefix' => null` to default options array on line 162.

**Comment by markstory** [comment]
@josegonzalez I added some tests around URL generation in the helper.

**Comment by markstory** [comment]
I'm a little concerned that `prefix` overlaps with routing prefixes in a confusing way. Some alternatives I thought of are:

* `queryStringPrefix`
* `parameterNamespace`
* `queryNamespace`
* `requestScope`

Do any of these (or another option) work better for people?

**Comment by htstudios** [comment]
I like requestScope the most, as it could possibly be applied to POST, too (say PRG).

**Comment by lorenzo** [comment]
`queryScope` ?

**Comment by ADmad** [comment]
Or just single worded `scope`?

**Comment by markstory** [comment]
@ADmad scope works too.

**Comment by lorenzo** [comment]
Tests failing :(

**Comment by ADmad** [comment]
The new scope option needs to be documented under relevant docblock.

**Comment by markstory** [comment]
Tests and API docs updated. :blush:

**Comment by htstudios** [comment]
Thank you guys!
This does mean that bake could now support multiple related pagination on view actions, right?

**Comment by dereuromark** [comment]
:+100:

**Pull request merged**